Fix monitoring

da-kami opened the pull request:
In this protocol we *can* actually have transactions with the same script, hence, we have to handle the `GetHistoryRes` by `Txid`.
To make this easier to understand we group the `Vec<Vec<GetHistoryRes>>` by `Txid` into `HashMap<Txid, GetHistroyRes>`.
Note: The assumption is, that only `GetHistroyRes` is returned for the `Txid`+`Script` combination.

fixes #260

da-kami committed:
Fix monitoring

In this protocol we *can* actually have transactions with the same script, hence, we have to handle the `GetHistoryRes` by `Txid`.
To make this easier to understand we group the `Vec<Vec<GetHistoryRes>>` by `Txid` into `HashMap<Txid, GetHistroyRes>`.
Note: The assumption is, that only `GetHistroyRes` is returned for the `Txid`+`Script` combination.

thomaseizinger commented on daemon/src/monitor.rs:
```diff
@@ -267,12 +267,37 @@ where
+                let txid = response.tx_hash;
+                let script = match txid_to_script.get(&txid) {
+                    None => {
+                        tracing::error!(
```
Use `warn` here? `error` seems a bit drastic.

thomaseizinger commented on daemon/src/monitor.rs:
```diff
@@ -298,30 +323,29 @@ where
+                let new_script_status = match histories.get(key) {
+                    None => ScriptStatus::Unseen,
+                    Some(history_entry) => {
+                        if history_entry.height <= 0 {
```
This could be a match arm guard.

thomaseizinger commented on daemon/src/monitor.rs:
```diff
@@ -298,30 +323,29 @@ where
+                        if history_entry.height <= 0 {
+                            ScriptStatus::InMempool
+                        } else {
+                            ScriptStatus::Confirmed(Confirmed::from_inclusion_and_latest_block(
```
As a clean-up, I think it would make sense to get rid of this `Confirmed` struct and just inline the entire logic into `ScriptStatus`.

thomaseizinger reviewed (approved):
Thanks for investigating and fixing this!

da-kami committed:
Cleanup `awaiting_status` in case we don't have remaining subscriptions

We always replaced it with `remaining`, but when populating `awaiting_status` upon sync we only go by the keys, so we actually fill in the script status again.
We should remove the complete entry if we want to make sure that we don't want to process it anymore.

Note: Kept seeing monitoring logs for things that were already past Finality (...) so we looked into this.

da-kami committed:
Avoid spamming the logs and add Todo

We keep seeing this in the logs, after a final tx went in:

```
2021-10-11 16:58:18 ERROR Could not find script in own state for txid b542cbc1920bdd926d79fb17377eda7152ed944ec64a1b09ce68dd11d9e7467e, ignoring
```

This happens because multiple transactions have the same script (e.g. CET, refund).
Once the CET is final we remove it from monitoring, which results in the txid not being available when matching.

thomaseizinger reviewed (approved):
Good work!

thomaseizinger commented on daemon/src/monitor.rs:
```diff
@@ -383,6 +383,12 @@ where
 
                     tracing::trace!("{} subscriptions reached their monitoring target, {} remaining for this script", reached_monitoring_target.len(), remaining.len());
 
+                    // TODO: When reaching finality of a final tx (CET, refund_tx,
+                    // collaborate_close_tx) we have to remove the remaining "competing"
+                    // transactions. This is not critical, but when fetching
+                    // `GetHistoryRes` by script we can have entries that we don't care about
+                    // anymore.
```
Make this an issue please.

da-kami replied:
https://github.com/comit-network/hermes/issues/276
